### PR TITLE
[PF-1585, PF-1586] Surface userFacingId instead of uuid

### PR DESCRIPTION
**Pull request by melissachang**
CLI commits are "released to prod" soon after they're merged. This PR uses `WorkspaceService.workspaceByUserFacingId`. **So I will wait for https://github.com/DataBiosphere/terra-workspace-manager/pull/636 to be in production (Broad and Verily) without any chance of rollback, before merging this PR.**

Before this PR, cli `ID` was `uuid`:
```
~/terra-cli (mc/ufid-create-update): terra workspace describe
ID: 66676d27-82e4-4f59-8f56-4202266c3258
Name:
Description:
Google project: terra-wsm-t-windy-truffle-3019
Cloud console: https://console.cloud.google.com/home/dashboard?project=terra-wsm-t-windy-truffle-3019
# Resources: 0
```

After this PR, `ID` is `userFacingId`:
```
~/terra-cli (mc/ufid-create-update): terra workspace describe
ID: user-facing-id
Name:
Description:
Google project: terra-wsm-t-windy-truffle-3019
Cloud console: https://console.cloud.google.com/home/dashboard?project=terra-wsm-t-windy-truffle-3019
# Resources: 0
```

This PR updates all CLI output, not just `terra workspace describe`.

Manual testing:
```
~/terra-cli (mc/ufid-create-update): terra workspace create --name="yo 1"
Workspace successfully created.
ID: a-0a942dea-f4be-4e3a-ba8c-c779d6c920d6
Name: yo 1
Description:
Google project: terra-wsm-t-silky-grape-2
Cloud console: https://console.cloud.google.com/home/dashboard?project=terra-wsm-t-silky-grape-2
# Resources: 0
~/terra-cli (mc/ufid-create-update): terra workspace create --name="yo 2"
Workspace successfully created.
ID: a-e601d42b-71e2-4d7f-a601-08599396e308
Name: yo 2
Description:
Google project: terra-wsm-t-sleek-walnut-9935
Cloud console: https://console.cloud.google.com/home/dashboard?project=terra-wsm-t-sleek-walnut-9935
# Resources: 0
~/terra-cli (mc/ufid-create-update): terra workspace list
   ID                                    NAME                            GOOGLE PROJECT                  DESCRIPTION
   a-0a942dea-f4be-4e3a-ba8c-c779d6c...  yo 1                            terra-wsm-t-silky-grape-2       (unset)
 ✓ a-e601d42b-71e2-4d7f-a601-0859939...  yo 2                            terra-wsm-t-sleek-walnut-9935   (unset)
~/terra-cli (mc/ufid-create-update): terra workspace list --format=json

[ {
  "id" : "a-0a942dea-f4be-4e3a-ba8c-c779d6c920d6",
  "name" : "yo 1",
  "description" : "",
  "googleProjectId" : "terra-wsm-t-silky-grape-2",
  "serverName" : "broad-dev-local-wsm",
  "userEmail" : "melchang@google.com"
}, {
  "id" : "a-e601d42b-71e2-4d7f-a601-08599396e308",
  "name" : "yo 2",
  "description" : "",
  "googleProjectId" : "terra-wsm-t-sleek-walnut-9935",
  "serverName" : "broad-dev-local-wsm",
  "userEmail" : "melchang@google.com"
} ]
~/terra-cli (mc/ufid-create-update):
~/terra-cli (mc/ufid-create-update): terra config get workspace

a-e601d42b-71e2-4d7f-a601-08599396e308
~/terra-cli (mc/ufid-create-update):
~/terra-cli (mc/ufid-create-update): terra workspace describe

ID: a-e601d42b-71e2-4d7f-a601-08599396e308
Name: yo 2
Description:
Google project: terra-wsm-t-sleek-walnut-9935
Cloud console: https://console.cloud.google.com/home/dashboard?project=terra-wsm-t-sleek-walnut-9935
# Resources: 0
~/terra-cli (mc/ufid-create-update):
~/terra-cli (mc/ufid-create-update): terra resource list

NAME                            RESOURCE TYPE         STEWARDSHIP TYPE      DESCRIPTION

~/terra-cli (mc/ufid-create-update):
~/terra-cli (mc/ufid-create-update): terra resource create gcs-bucket --name=yo_2

Successfully added controlled GCS bucket.
Name:         yo_2
Description:
Type:         GCS_BUCKET
Stewardship:  CONTROLLED
Cloning:      COPY_RESOURCE
Access scope: SHARED_ACCESS
Managed by:   USER
GCS bucket name: terra-0be639d0-b70b-4f50-8398-3ba6ae7940f9-bucket
Location: US-CENTRAL1
# Objects: 0
~/terra-cli (mc/ufid-create-update):
~/terra-cli (mc/ufid-create-update): terra resource list
NAME                            RESOURCE TYPE         STEWARDSHIP TYPE      DESCRIPTION
yo_2                            GCS_BUCKET            CONTROLLED            (unset)
~/terra-cli (mc/ufid-create-update): terra resource list --workspace=a-e601d42b-71e2-4d7f-a601-08599396e308
NAME                            RESOURCE TYPE         STEWARDSHIP TYPE      DESCRIPTION
yo_2                            GCS_BUCKET            CONTROLLED            (unset)
~/terra-cli (mc/ufid-create-update): terra resource create gcs-bucket --name=yo_1 --workspace=a-0a942dea-f4be-4e3a-ba8c-c779d6c920d6
Successfully added controlled GCS bucket.
Name:         yo_1
Description:
Type:         GCS_BUCKET
Stewardship:  CONTROLLED
Cloning:      COPY_RESOURCE
Access scope: SHARED_ACCESS
Managed by:   USER
GCS bucket name: terra-2d120658-75bf-4b90-b6c8-1ac3a48da311-bucket
Location: US-CENTRAL1
# Objects: 0
~/terra-cli (mc/ufid-create-update): terra resource list --workspace=a-0a942dea-f4be-4e3a-ba8c-c779d6c920d6
NAME                            RESOURCE TYPE         STEWARDSHIP TYPE      DESCRIPTION
yo_1                            GCS_BUCKET            CONTROLLED            (unset)
~/terra-cli (mc/ufid-create-update): terra workspace update --name="yoyo 2"
Workspace successfully updated.
ID: a-e601d42b-71e2-4d7f-a601-08599396e308
Name: yoyo 2
Description:
Google project: terra-wsm-t-sleek-walnut-9935
Cloud console: https://console.cloud.google.com/home/dashboard?project=terra-wsm-t-sleek-walnut-9935
# Resources: 0
~/terra-cli (mc/ufid-create-update): terra workspace describe
ID: a-e601d42b-71e2-4d7f-a601-08599396e308
Name: yoyo 2
Description:
Google project: terra-wsm-t-sleek-walnut-9935
Cloud console: https://console.cloud.google.com/home/dashboard?project=terra-wsm-t-sleek-walnut-9935
# Resources: 0
~/terra-cli (mc/ufid-create-update): terra workspace list
   ID                                    NAME                            GOOGLE PROJECT                  DESCRIPTION
   a-0a942dea-f4be-4e3a-ba8c-c779d6c...  yo 1                            terra-wsm-t-silky-grape-2       (unset)
 ✓ a-e601d42b-71e2-4d7f-a601-0859939...  yoyo 2                          terra-wsm-t-sleek-walnut-9935   (unset)
~/terra-cli (mc/ufid-create-update): terra workspace clone
# Ignore error, this is due to PF-1623
Resource not found: yo_2
~/terra-cli (mc/ufid-create-update):
~/terra-cli (mc/ufid-create-update): terra workspace list
   ID                                    NAME                            GOOGLE PROJECT                  DESCRIPTION
   a-02f63a7b-1f5f-4772-8f2b-b79f940...  (unset)                         terra-wsm-t-lucent-acorn-9463   (unset)
   a-0a942dea-f4be-4e3a-ba8c-c779d6c...  yo 1                            terra-wsm-t-silky-grape-2       (unset)
 ✓ a-e601d42b-71e2-4d7f-a601-0859939...  yoyo 2                          terra-wsm-t-sleek-walnut-9935   (unset)
~/terra-cli (mc/ufid-create-update): terra workspace delete --quiet
ID: a-e601d42b-71e2-4d7f-a601-08599396e308
Name: yoyo 2
Description:
Google project: terra-wsm-t-sleek-walnut-9935
Cloud console: https://console.cloud.google.com/home/dashboard?project=terra-wsm-t-sleek-walnut-9935
# Resources: 0
Workspace successfully deleted.
~/terra-cli (mc/ufid-create-update): terra workspace list
   ID                                    NAME                            GOOGLE PROJECT                  DESCRIPTION
   a-02f63a7b-1f5f-4772-8f2b-b79f940...  (unset)                         terra-wsm-t-lucent-acorn-9463   (unset)
   a-0a942dea-f4be-4e3a-ba8c-c779d6c...  yo 1                            terra-wsm-t-silky-grape-2       (unset)
```

Ignore failing test `WorkspaceSetDeferLogin`. It will be deleted in PF-1622.

Next PRs:
- Allow setting ufId with create, update, clone
- Require ufId on create, clone

FYI @ginaygoogle

### CODE REVIEW

**Comment by jaycarlton** [comment]
Can we keep the workspace UUID in the JSON output for old times' sake (and for debugging)? There _might_ be situations where power users scripting the CLI will want an immutable identifier. I don't think it would confuse folks much if you named it `uuid` in the JSON.

File: src/main/java/bio/terra/cli/businessobject/Workspace.java [hunk 1]
```diff
@@ -164,6 +159,19 @@ public static Workspace get(UUID id) {
...
   }
 
+  /** Fetch an existing workspace by userFacingId, with resources populated */
+  public static Workspace get(String userFacingId) {
+    // call WSM to fetch the existing workspace object and backing Google context
+    WorkspaceDescription loadedWorkspace =
+        WorkspaceManagerService.fromContext().getWorkspaceByUserFacingId(userFacingId);
+    logger.info("Loaded workspace: {}", loadedWorkspace);
+
+    // convert the WSM object to a CLI object
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/businessobject/Workspace.java:169]: this bit is shared with the overload for `get()`. You could pull out a `getPopulatedWorksapce(WorkspaceDescription desc)` and use it in both places.

File: src/main/java/bio/terra/cli/businessobject/Workspace.java [hunk 2]
```diff
@@ -331,8 +339,12 @@ public String grantBreakGlass(
 
   // ====================================================
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/businessobject/Workspace.java:340]: Can we get rid of this decoration? I know what accessors and mutators are.

File: src/main/java/bio/terra/cli/businessobject/WorkspaceUser.java [hunk 3]
```diff
@@ -93,12 +94,12 @@ public static WorkspaceUser add(String email, Role role, Workspace workspace) {
   public static WorkspaceUser remove(String email, Role role, Workspace workspace) {
     // call WSM to remove a user + role from the current workspace
     WorkspaceManagerService.fromContext()
-        .removeIamRole(workspace.getId(), email, role.getWsmRole());
+        .removeIamRole(workspace.getUuid(), email, role.getWsmRole());
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/businessobject/WorkspaceUser.java:97]: This ship may have sailed, but could you make a separate PR for the automatic rename `id -> uuid` and kin? It's easier to approve them if the developer asserts it was an automatic refactoring.

File: src/main/java/bio/terra/cli/command/shared/options/WorkspaceOverride.java [hunk 4]
```diff
@@ -18,12 +17,12 @@ public class WorkspaceOverride {
...
       names = "--workspace",
       description = "Workspace id to use for this command only.")
-  private UUID id;
+  private String userFacingId;
 
   /** Helper method to override the current workspace if the `--workspace` flag specifies an id. */
   public void overrideIfSpecified() {
-    if (id != null) {
-      Context.useOverrideWorkspace(id);
+    if (userFacingId != null) {
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/command/shared/options/WorkspaceOverride.java:24]: maybe check for empty string here?

File: src/main/java/bio/terra/cli/command/workspace/List.java [hunk 5]
```diff
@@ -68,9 +70,9 @@ private void printText(java.util.List<UFWorkspaceLight> returnValue) {
 
   /** Column information for table output with `terra workspace list` */
   private enum Columns implements ColumnDefinition<UFWorkspaceLight> {
+    ID("ID", w -> w.id, 36, RIGHT),
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/command/workspace/List.java:73]: You might want to increase this column's width at the expense of the description or display name, since it's the ID we'll be using to do things.

File: src/main/java/bio/terra/cli/serialization/persisted/PDWorkspace.java [hunk 6]
```diff
@@ -28,7 +29,8 @@ public class PDWorkspace {
 
   /** Serialize an instance of the internal class to the disk format. */
   public PDWorkspace(Workspace internalObj) {
-    this.id = internalObj.getId();
+    this.uuid = internalObj.getUuid();
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/serialization/persisted/PDWorkspace.java:32]: Are you attempting backward compatibility? This will break that. I'd be inclined to keep the `id` since it's a JSON field.

File: src/main/java/bio/terra/cli/serialization/userfacing/UFConfig.java [hunk 7]
```diff
@@ -91,7 +91,8 @@ public static class Builder {
     private Logger.LogLevel fileLoggingLevel;
     private Logger.LogLevel consoleLoggingLevel;
     private String serverName;
-    private UUID workspaceId;
+    // "id" instead of "userFacingId" because user sees this with "terra config list"
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/serialization/userfacing/UFConfig.java:94]: I'm confused here. I thought you were keeping both IDs in the config.

File: src/main/java/bio/terra/cli/serialization/userfacing/UFWorkspace.java [hunk 8]
```diff
@@ -48,15 +47,15 @@ public void print() {
 
   @JsonPOJOBuilder(buildMethodName = "build", withPrefix = "")
   public static class Builder {
-    private UUID id;
+    private String id;
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/serialization/userfacing/UFWorkspace.java:50]: Let's keep both for the sake of JSON.

File: src/main/java/bio/terra/cli/service/WorkspaceManagerService.java [hunk 9]
```diff
@@ -298,21 +298,41 @@ public WorkspaceDescription createWorkspace(
...
-   *
-   * @param workspaceId the id of the workspace to fetch
-   * @return the Workspace Manager workspace description object
    */
-  public WorkspaceDescription getWorkspace(UUID workspaceId) {
+  public WorkspaceDescription getWorkspace(UUID uuid) {
+    WorkspaceDescription workspaceWithContext =
+        callWithRetries(
+            () -> new WorkspaceApi(apiClient).getWorkspace(uuid), "Error fetching workspace");
+    String googleProjectId =
```
Comment (jaycarlton) [inline src/main/java/bio/terra/cli/service/WorkspaceManagerService.java:306]: You can DRY these two functions.

File: src/test/java/unit/BqDatasetControlled.java [hunk 10]
```diff
@@ -36,7 +35,7 @@ void listDescribeReflectCreate() throws IOException {
     // `terra workspace set --id=$id --format=json`
     UFWorkspace workspace =
         TestCommand.runAndParseCommandExpectSuccess(
-            UFWorkspace.class, "workspace", "set", "--id=" + getWorkspaceId());
+            UFWorkspace.class, "workspace", "set", "--id=" + getUserFacingId());
```
Comment (jaycarlton) [inline src/test/java/unit/BqDatasetControlled.java:38]: Are you 100% sure it's OK to break existing scripts that use `id` to mean `uuid`? Maybe send Nicole some flowers.

**Review by jaycarlton — APPROVED** [review]
Looks great. Do think about exposing the UUIDs for power users somehow. Easiest way is with a `uuid` field in the JSON.

File: src/test/java/unit/BqDatasetControlled.java (hunk 10 above)
Comment (cbookg) [reply src/test/java/unit/BqDatasetControlled.java:38]: Maybe this flag should accept either UUID or UFID?

File: src/main/java/bio/terra/cli/serialization/persisted/PDWorkspace.java (hunk 6 above)
Comment (cbookg) [reply src/main/java/bio/terra/cli/serialization/persisted/PDWorkspace.java:32]: Maybe okay to break compatibility now.  But... maybe we want to store UUID on the disk as the "current workspace"?

File: src/test/java/unit/BqDatasetControlled.java (hunk 10 above)
Comment (jaycarlton) [reply src/test/java/unit/BqDatasetControlled.java:38]: If we had existing users, I would've kept `id` to mean uuid and used `ufid` to mean the new ID, to prevent breakage. I agree we need a way to specify the `uuid` in certain scenarios (like a user has a spreadsheet with a row for each workspace and needs a unique immutable key to update it).

File: src/main/java/bio/terra/cli/businessobject/Workspace.java (hunk 1 above)
Comment (melissachang) [reply src/main/java/bio/terra/cli/businessobject/Workspace.java:169]: Done

File: src/main/java/bio/terra/cli/businessobject/Workspace.java (hunk 2 above)
Comment (melissachang) [reply src/main/java/bio/terra/cli/businessobject/Workspace.java:340]: Done

File: src/main/java/bio/terra/cli/businessobject/WorkspaceUser.java (hunk 3 above)
Comment (melissachang) [reply src/main/java/bio/terra/cli/businessobject/WorkspaceUser.java:97]: Excellent idea, will do next time :)

File: src/main/java/bio/terra/cli/command/shared/options/WorkspaceOverride.java (hunk 4 above)
Comment (melissachang) [reply src/main/java/bio/terra/cli/command/shared/options/WorkspaceOverride.java:24]: Done

File: src/main/java/bio/terra/cli/command/workspace/List.java (hunk 5 above)
Comment (melissachang) [reply src/main/java/bio/terra/cli/command/workspace/List.java:73]: My thinking is that most IDs will be less than 36, and won't be cut off.

Right now terminal has to be 145 characters wide (otherwise we wrap):

image

Maybe we should make Description shorter?

File: src/main/java/bio/terra/cli/serialization/userfacing/UFConfig.java (hunk 7 above)
Comment (melissachang) [reply src/main/java/bio/terra/cli/serialization/userfacing/UFConfig.java:94]: I'm keeping both in `PDWorkspace`. `PD` is physical disk -- what we're storing in `context.json` file. We need `uuid` there because WSM APIs take `uuid`, not `userFacingId`.

`UF` in `UFConfig` refers to what users see when they run a CLI command. I'd like to not show `uuid` there.

**Review by melissachang — COMMENTED** [review]
> Can we keep the workspace UUID in the JSON output for old times' sake (and for debugging)? There _might_ be situations where power users scripting the CLI will want an immutable identifier. I don't think it would confuse folks much if you named it `uuid` in the JSON.

I forgot to mention, `uuid` is in `context.json` file that's stored on disk:

```
  "workspace" : {
    "uuid" : "0152a77f-0f0c-430b-94ab-b4798fa5a990",
    "userFacingId" : "a0152a77f-0f0c-430b-94ab-b4798fa5a990",
```

Also in `~/.terra/logs/terra.log`.

Could that suffice for the power user? I don't think we have to worry about backwards compatibility, considering we don't have external users yet. (I will let solutions folks know.)

How about we don't expose `uuid` for now (other than `context.json`, `terra.log`), and if a customer asks for this, we can revisit?

Btw I [added some documentation.](https://github.com/DataBiosphere/terra-cli/blob/a31678cac4da88a17caa67ca074b01dd768d9df1/CONTRIBUTING.md#workspace-ids)

**Review by cbookg — APPROVED** [review]

**Comment by melissachang** [comment]
@zloery After rebasing #246, I changed [some code](https://github.com/DataBiosphere/terra-cli/pull/238/commits/cd714ef2f4bfcb13bf053500b6d2adf2e7be51ba#diff-30aa582ff9a993e1fd074b23c98db11b6c01b44370a46137e19d3062cea5a65f) to get UUID from `context.json` instead of `UFWorkspace`.
